multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedSetCache.java

```diff
@@ -44,8 +44,18 @@ public EmbeddedSetCache(Cache<K, SetBucket<V>> cache) {
     * @param key, the name of the set
     * @return the set with values if such exist, or null if the key is not present
     */
-   public CompletionStage<Set<V>> get(K key) {
-      return readWriteMap.eval(key, new SGetFunction<K, V>()).thenApply(v -> v.values());
+   public CompletionStage<SetBucket<V>> get(K key) {
```

Comment:
We should really be doing this with new functions 
1. atomic remove which returns if the value was removed 
2. atomic add that adds the element to a set

Technically this won't make the SMOVE operation itself atomic, but at least the updates to the sets themselves can be atomic.

Reply:
do you mean remove and add in two steps?
my intention was to do a single putAll to ensure both sets are updated in the same operation... but may this doesn't guarantee atomicity

Reply:
+1 to using functions, too. Which also helps a bit in transferring the entry back and forth between nodes.

I think we can't provide an atomic operation as we don't load and lock both entries. There is always a small window when the entry will be neither on source nor destination.

Reply:
yeah ok. I'll follow the lmove approach

Reply:
changed the implementation to functions